### PR TITLE
1122: Compressed negative slices list

### LINKED ISSUE #1122: List of bad slices in error message could be compressed
### Current Behaviour

If the output of flat-fielding has negative values then a complete list problem slices is written to the log.

### Desired Behaviour

Changes this write a more compact message, e.g
Negative values found on slices 234, 456
Negative values found on all slices
Negative values found on slices 1-25, 30, 60-100

The message could then be displayed in full in the GUI.

### PR DESCRIPTION
**Pull request by DolicaAkelloEgwel**
### Issue

Closes #1122

### Description

Simplifies the list of slices with negative values after flat fielding.

### Testing 

Updated an existing test to check that the log message contains ranges "0-2, 4-5, 7" rather than printing the index of every slice with a negative value. Also added a test for an "all slices" are message if all slices contain negative values.

### Acceptance Criteria 

Check that the tests pass. Check that the slice range is displayed in the log message after running flat-fielding.

### Documentation

Updated release notes.

### CODE REVIEW

File: mantidimaging/gui/windows/operations/presenter.py [hunk 1]
```diff
@@ -54,6 +60,33 @@ def _find_nan_change(before_image, filtered_image_data):
...
+def sub(x: Tuple[int, int]) -> int:
+    """
+    Subtracts two tuples. Helper method for generating the negative slice list.
+    :param x: The int tuple.
+    :return: The first value subtracted from the second value.
+    """
+    return x[1] - x[0]
+
+
+def _generate_slices_range_list(slices: List[int]) -> List[str]:
```
Comment (samtygier-stfc) [inline mantidimaging/gui/windows/operations/presenter.py:72]: Could you rename this to _group_consecutive_values() to make it clearer what it does.

And also add a simple unit test just for this function.

File: mantidimaging/gui/windows/operations/presenter.py [hunk 2]
```diff
@@ -401,7 +430,12 @@ def _show_negative_values_error(self, negative_stacks: List[StackVisualiserView]
             for i in range(len(stack.presenter.images.data)):
                 if np.any(stack.presenter.images.data[i] < 0):
                     negative_slices.append(i)
-            getLogger(__name__).error(f"Slices containing negative values in {stack.name}: {negative_slices}")
+            slices_msg = f'Slices containing negative values in {stack.name}: '
+            if len(negative_slices) == len(stack.presenter.images.data):
+                slices_msg += "all slices."
+            else:
+                slices_msg += f'{", ".join(_generate_slices_range_list(negative_slices))}'
+            getLogger(__name__).error(slices_msg)
```
Comment (samtygier-stfc) [inline mantidimaging/gui/windows/operations/presenter.py:438]: Could you use the same message in both the GUI and the log message.

Maybe to avoid the GUI message getting to long something like
```
ranges = _generate_slices_range_list(negative_slices)
if len(ranges) > 12:
  message = "  {", ".join(ranges[:10])} ... ranges[-1]" 
else:
  message = "  {", ".join(ranges)}" 
```

File: mantidimaging/gui/windows/operations/test/test_presenter.py [hunk 3]
```diff
@@ -422,13 +422,41 @@ def test_negative_values_in_flat_fielding_shows_error(self, do_apply_filter_sync
         self.presenter.view.filterSelector.currentText.return_value = FLAT_FIELDING
         mock_task = mock.Mock()
         mock_task.error = None
-        self.mock_stack_visualisers[0].presenter.images.data = np.array([-1 for _ in range(3)])
+        images = generate_images()
+        images.data[0][0][0] = -1
```
Comment (samtygier-stfc) [inline mantidimaging/gui/windows/operations/test/test_presenter.py:426]: Can this be `images.data[0,0,0] = -1`

File: mantidimaging/gui/windows/operations/test/test_presenter.py [hunk 4]
```diff
@@ -422,13 +422,41 @@ def test_negative_values_in_flat_fielding_shows_error(self, do_apply_filter_sync
...
+
+    @mock.patch('mantidimaging.gui.windows.operations.presenter.FiltersWindowPresenter._do_apply_filter_sync')
+    def test_negative_values_in_all_slices(self, do_apply_filter_sync_mock):
+        self.presenter.view.safeApply.isChecked.return_value = False
+        self.presenter.view.filterSelector.currentText.return_value = FLAT_FIELDING
+        mock_task = mock.Mock()
+        mock_task.error = None
+        images = generate_images()
+        for i in range(len(images.data)):
+            images.data[i][0][0] = -1
```
Comment (samtygier-stfc) [inline mantidimaging/gui/windows/operations/test/test_presenter.py:452]: Can this be
`images.data[:,0,0] = -1`

**Review by samtygier-stfc — CHANGES_REQUESTED** [review]
Nice approach. Just a few tweaks.

**Comment by samtygier-stfc** [comment]
Looks good :+1: 
image

I put the following in _show_negative_values_error() to see how it looked
```
import random
negative_slices = [x for x in range(len(stack.presenter.images.data)) if random.random() > 0.5]
```

**Review by samtygier-stfc — APPROVED** [review]
